_posts/2017-04-22-Annoying-Mastodon-Users-with-a-Bot.md

```diff
@@ -0,0 +1,212 @@
+end
+```
+
+Now we'll add some methods to handle the interaction with the Microsoft
+Translator API.
+
+```ruby
+# Gets an access token for the Translator API
+def get_token(translator_token)
+  token_url = "https://api.cognitive.microsoft.com/sts/v1.0/issueToken"
+  response = Curl.post(token_url) do |response|
+    response.headers['Ocp-Apim-Subscription-key'] = translator_token
+  end
+
+  return response.body_str
+end
+
+# Connects to the API and translate the toot's content.
+# If there's an issue with the connection, we return `false`
+def translate_toot(toot, token)
+  toot_content = toot.content.without_tags
+  translation_url = 
+    "http://api.microsofttranslator.com/v2/Http.svc/Translate?" +
+    "text=" + toot_content + "&to=en&from=ja"
+
+  begin
+    response = Curl.get(translation_url) do |response|
+      response.headers['Authorization'] = "Bearer " + token
+    end
+  rescue
+    return false
+  end
+
+  if response.status != "200 OK"
+    return false
+  end
+
+  return response.body_str.without_tags
+end
+```
+
+With that in place, we can setup our basic program logic.
+It needs to:
+- Connect to Mastodon and grab some toots to chew on
+- Decide which toots are in Japanese
+- Connect to Microsoft and get a fresh access token
+- Translate the Japanese toots
+- Reply to the Japanese toots with the Translator API translation.
+
+```ruby
+# Connect to mastodon
+mastodon = Mastodon::REST::Client.new(
+  base_url: 'https://mastodon.cloud', bearer_token: mastodon_token
+)
+
+# Grab the 25 most recent toots from the Public Timeline
+toots = mastodon.public_timeline(limit:25)
+
+# Find all of the Japanese toots and translate them
+japanese_toots = []
+token = get_token(translator_token)
+toots.each do |toot|
+  toot_content = toot.content.without_tags
+  if toot_content.japanese?
+    translated = translate_toot(
+      toot, token
+    )
+    if translated
+      japanese_toots << [toot, translated]
+    end
+  end
+end
+
+# Loop through the Japanese toots and start replying
+japanese_toots.each do |toot|
+  begin
+    # toot[1] is the translated content of the toot
+    translated_toot = "English Translation:\n" + toot[1]
+
+    # Make sure we don't go over the maximum toot length
+    if translated_toot.length <= 500
+      # toot[0] is the original toot we are replying to
+      mastodon.create_status(
+        translated_toot,
+        toot[0].id
+      )
+      puts "Tooted @ " + toot[0].account.acct
+    end
+  rescue
+    puts "Failed tooting!"
+  end
+end
+```
+
+You can see the completed script in [this gist][gist].
+
+I wrote a slightly different version of this script that bulls in the tokens
+from a JSON file, and also continuously loops to toot out replies to new
+statuses as they come in. If you'd like to take a look at that script, check
+it out in [this repo][translator-bot].
+
+### In action
+
+I ran this bot for a few days under the handle
+[@japanesetranslator][translator]. It seemed to annoy quite a few instance
```

Comment:
I would provide a few links to annoyed users or instance admins, like:
>It [seemed to annoy](#) [quite a few](#) [instance admins](#), since it sent out […]
Note that there are multiple dummy links in this example.